### PR TITLE
Bounding Box Tests

### PR DESCRIPTION
**Pull request by julenka**
## Overview
Add bounding box tests, a couple more small test utilities, and simulated hand ray not showing up until you move hand a bit.

## Changes
- Fixes: #4730 by implementing all tests outlined in the bug. The tests actually highlighted bug #5029 which I then fixed thanks to @luis-valverde-ms excellent but description
- Fixes: #5029 
- Add couple utility methods to tests
- Make a couple members in input simulation service set-table to make them easier to test.

Thanks to Lukas for making the input simulation system easy to modify at run time so I can do near and far HoloLens 1 style tests :-)

## Verification
- PlayMode tests pass individually
- When I run all the tests, I get a stale scene left over, need to fix still.
- Getting the following error currently when I run any test (whether it is the tests I added, or not). Possibly existing bug:

```
NullReferenceException: Object reference not set to an instance of an object
UnityEditor.TestTools.TestRunner.PlaymodeLauncher+BackgroundWatcher.OnPlayModeStateChanged
```

### CODE REVIEW

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+            bbox.transform.localScale *= 0.5f;
+            bbox.Active = true;
+
+            return bbox;
+        }
+        #endregion
+
+        [UnityTest]
+        public IEnumerator BBoxInstantiate()
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs]: add comment

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+        #region Utilities
+        [TearDown]
+        public void ShutdownMrtk()
+        {
+            TestUtilities.ShutdownMixedRealityToolkit();
+        }
+
+        /// <summary>
+        /// Instantiates a bounding box at 0, 0, -1.5f
+        /// box is at scale 1,1,1
```
Comment (julenka) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs]: scale is 0.5, fix comment

File: Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs [hunk 3]
```diff
@@ -12,7 +12,7 @@ namespace Microsoft.MixedReality.Toolkit.Input
     /// </summary>
     public abstract class InputSystemGlobalHandlerListener : MonoBehaviour
     {
-        private bool lateInitialize = true;
+        private bool lateInitialize = false;
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs]: @googlan Is this a bugfix? Doesn't look related to the rest of the changes here.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+            var newObject = new GameObject();
+            var bc = newObject.AddComponent<BoxCollider>();
+            bc.center = new Vector3(.25f, 0, 0);
+            bc.size = new Vector3(0.162f, 0.1f, 1);
+            bbox.BoundsOverride = bc;
+            yield return null;
+
+            // Get the dimensions of the corners
+            List<GameObject> corners = FindDescendantsContainingName(bbox.gameObject, "corner_");
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs]: Can we add a function to BoundingBox to get the GameObject for all or one specific corner(s)? Could just be a wrapper around the FindDescendants, but would keep the logic inside BoundingBox.cs.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+            TestUtilities.AssertAboutEqual(bounds.center, startCenter, "bbox incorrect center at start");
+            TestUtilities.AssertAboutEqual(bounds.size, startSize, "bbox incorrect size at start");
+
+            var iss = PlayModeTestUtilities.GetInputSimulationService();
+            var oldIsp = iss.InputSimulationProfile;
+            var isp = new MixedRealityInputSimulationProfile
+            {
+                HandSimulationMode = HandSimulationMode.Gestures
+            };
+            iss.InputSimulationProfile = isp;
```
Comment (lukastoenneMS) [inline Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs]: So swapping out just the input sim profile works like this? I tried something similar for writing gesture event tests, but was setting the entire MRTK.Instance.ActiveProfile because i didn't want to risk changing assets (the default input system profile) just by running tests. If the tests fail they would not reset the profile and inadvertently change the asset, which might get saved automatically by Unity. But changing the ActiveProfile at runtime causes a lot of broken behavior. Just changing the input sim profile probably is easier to keep working.

File: Assets/MixedRealityToolkit.Tests/TestUtilities.cs [hunk 6]
```diff
@@ -163,5 +165,11 @@ public static void ShutdownMixedRealityToolkit()
             return ScriptableObject.CreateInstance<T>();
 #endif
         }
+
+        public static void AssertAboutEqual(Vector3 actual, Vector3 expected, string message)
```
Comment (luis-valverde-ms) [inline Assets/MixedRealityToolkit.Tests/TestUtilities.cs]: Unity has something for this, have a look at Vector3EqualityComparer here: https://docs.unity3d.com/2017.4/Documentation/Manual/PlaymodeTestFramework.html.

Comment (thalbern) [reply Assets/MixedRealityToolkit.Tests/TestUtilities.cs]: thanks! I should use that for manipulationhandler as well - doing the manual magnitude compare there as well

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs (hunk 5 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs]: Yep, this was working for me! The only thing I needed to make sure was to change the profile back to original at the end of the test.

File: Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs (hunk 3 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs]: Yeah it's a bugfix, for #5029. I need it for the tests to pass. I can make it as a separate PR, but didn't want to add extra overhead of getting approve for such a small change.

Comment (julenka) [reply Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs]: I found the commit that introduced the lateInitialize variable:

```
1c17f7847e1d3a3a8e125b876a9c57ad1fb42a6f
Fixed input handler race condition if input system and teleport system not found before gameobjects wake up
```

I think with my change, the handler will never register itself if the input system is not available when the component gets enabled. I think we need one more change. I'll go ahead and pull that as separate pr.

File: Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs (hunk 4 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Tests/PlayModeTests/BoundingBoxTests.cs]: ah great suggestion, thanks.

**Comment by julenka** [comment]
This PR blocked by #5044 need to complete that first or tests won't pass

File: Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs (hunk 3 above)
Comment (julenka) [reply Assets/MixedRealityToolkit.Services/InputSystem/InputSystemGlobalHandlerListener.cs]: Moved to separate PR.

**Comment by julenka** [comment]
Error from CI is:

  F:\aipmragent_work\1\s\Assets\MixedRealityToolkit.Tests\TestUtilities.cs(8,19): **error CS0234: The type or namespace name 'TestTools' does not exist in the namespace 'UnityEngine' (are you missing an assembly reference?)** [f:\aipmragent_work\1\a\build\UWP_x86_.NET\GeneratedProjects\UWP\Microsoft.MixedReality.Toolkit.Tests\Microsoft.MixedReality.Toolkit.Tests.csproj]

Probably need to modify .asmdef files. Investigating.

**Comment by julenka** [comment]
WaitForEndOfFrame [is not invoked when running tests in batch mode](https://docs.unity3d.com/ScriptReference/WaitForEndOfFrame.html)
